runtime/lua/vim/lsp/buf.lua

```diff
@@ -270,58 +297,79 @@ local sig_help_ns = api.nvim_create_namespace('vim_lsp_signature_help')
 function M.signature_help(config)
   local method = ms.textDocument_signatureHelp
 
-  config = config or {}
+  config = config and vim.deepcopy(config) or {}
   config.focus_id = method
 
-  lsp.buf_request(0, method, client_positional_params(), function(err, result, ctx)
-    local client = assert(vim.lsp.get_client_by_id(ctx.client_id))
-
-    if err then
-      vim.notify(
-        client.name .. ': ' .. tostring(err.code) .. ': ' .. err.message,
-        vim.log.levels.ERROR
-      )
-      api.nvim_command('redraw')
-      return
-    end
-
+  lsp.buf_request_all(0, method, client_positional_params(), function(results, ctx)
     if api.nvim_get_current_buf() ~= ctx.bufnr then
       -- Ignore result since buffer changed. This happens for slow language servers.
       return
     end
 
-    -- When use `autocmd CompleteDone <silent><buffer> lua vim.lsp.buf.signature_help()` to call signatureHelp handler
-    -- If the completion item doesn't have signatures It will make noise. Change to use `print` that can use `<silent>` to ignore
-    if not result or not result.signatures or not result.signatures[1] then
+    local signatures = process_signature_help_results(results)
+
+    if not next(signatures) then
       if config.silent ~= true then
         print('No signature help available')
       end
       return
     end
 
-    local triggers =
-      vim.tbl_get(client.server_capabilities, 'signatureHelpProvider', 'triggerCharacters')
-
     local ft = vim.bo[ctx.bufnr].filetype
-    local lines, hl = util.convert_signature_help_to_markdown_lines(result, ft, triggers)
-    if not lines or vim.tbl_isempty(lines) then
-      if config.silent ~= true then
-        print('No signature help available')
+    local total = #signatures
+    local idx = 0
+
+    --- @param update_win? integer
+    local function show_signature(update_win)
+      idx = (idx % total) + 1
+      local client, result = signatures[idx][1], signatures[idx][2]
+      --- @type string[]?
+      local triggers =
+        vim.tbl_get(client.server_capabilities, 'signatureHelpProvider', 'triggerCharacters')
+      local lines, hl =
+        util.convert_signature_help_to_markdown_lines({ signatures = { result } }, ft, triggers)
+      if not lines then
+        return
       end
-      return
+
+      local sfx = total > 1 and string.format(' (%d/%d) (<C-s> to cycle)', idx, total) or ''
+      local title = string.format('Signature Help: %s%s', client.name, sfx)
+      if config.border then
+        config.title = title
+      else
+        table.insert(lines, 1, '# ' .. title)
+        if hl then
+          hl[1] = hl[1] + 1
+          hl[3] = hl[3] + 1
+        end
+      end
+
+      config._update_win = update_win
+
+      local buf, win = util.open_floating_preview(lines, 'markdown', config)
```

Comment:
Can we just hardcode `markdown` here? What if a client sends documentation as plain text?

Reply:
I'm not sure. This is what was here before.

Reply:
`lines` is explicitly provided as markdown from `vim.lsp.util.convert_signature_help_to_markdown_lines()`